### ISSUE TITLE
Bcachefs having poor disk random read performance

### ISSUE THREAD

**Issue by bhzhu203** [koverstreet/bcachefs#387]
Bcachefs version 62cdb94d54b37e9e1fb8c73d1fe2b8ab35a88d2f . The read iops far less than the write iops.

```
❰bhzhu❙~❱✘≻ fio -ioengine=libaio -bs=4k -direct=1 -thread -rw=randread -size=1G -filename=test1 -name="Max throughput" -iodepth=4 -runtime=60
Max throughput: (g=0): rw=randread, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=4
fio-3.28
Starting 1 thread
Max throughput: Laying out IO file (1 file / 1024MiB)
Jobs: 1 (f=1): [r(1)][100.0%][r=920KiB/s][r=230 IOPS][eta 00m:00s]
Max throughput: (groupid=0, jobs=1): err= 0: pid=200001: Fri Mar 18 19:25:23 2022
  read: IOPS=232, BW=930KiB/s (952kB/s)(54.5MiB/60021msec)
    slat (usec): min=6, max=699, avg=39.48, stdev=18.46
    clat (usec): min=91, max=147429, avg=17163.88, stdev=15790.82
     lat (usec): min=113, max=147463, avg=17203.79, stdev=15790.82
    clat percentiles (usec):
     |  1.00th=[   229],  5.00th=[  2835], 10.00th=[  4015], 20.00th=[  6128],
     | 30.00th=[  7701], 40.00th=[  9372], 50.00th=[ 11863], 60.00th=[ 14746],
     | 70.00th=[ 19006], 80.00th=[ 26084], 90.00th=[ 37487], 95.00th=[ 50070],
     | 99.00th=[ 76022], 99.50th=[ 86508], 99.90th=[116917], 99.95th=[125305],
     | 99.99th=[137364]
   bw (  KiB/s): min=  464, max= 1112, per=100.00%, avg=930.15, stdev=87.78, samples=119
   iops        : min=  116, max=  278, avg=232.54, stdev=21.95, samples=119
  lat (usec)   : 100=0.01%, 250=1.64%, 500=0.11%, 1000=0.01%
  lat (msec)   : 2=0.30%, 4=7.88%, 10=32.91%, 20=28.64%, 50=23.45%
  lat (msec)   : 100=4.80%, 250=0.23%
  cpu          : usr=0.18%, sys=1.09%, ctx=13945, majf=0, minf=5
  IO depths    : 1=0.1%, 2=0.1%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=13951,0,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=4

Run status group 0 (all jobs):
   READ: bw=930KiB/s (952kB/s), 930KiB/s-930KiB/s (952kB/s-952kB/s), io=54.5MiB (57.1MB), run=60021-60021msec

Disk stats (read/write):
  sdb: ios=14219/203, merge=81/58, ticks=262576/5195, in_queue=272279, util=99.66%
❰bhzhu❙~❱✔≻ fio -ioengine=libaio -bs=4k -direct=1 -thread -rw=randwrite -size=1G -filename=test1 -name="Max throughput" -iodepth=4 -runtime=60
Max throughput: (g=0): rw=randwrite, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=4
fio-3.28
Starting 1 thread
Jobs: 1 (f=1): [w(1)][96.6%][w=488KiB/s][w=122 IOPS][eta 00m:01s]   
Max throughput: (groupid=0, jobs=1): err= 0: pid=200221: Fri Mar 18 19:26:17 2022
  write: IOPS=9225, BW=36.0MiB/s (37.8MB/s)(1024MiB/28415msec); 0 zone resets
    slat (usec): min=2, max=15365, avg= 8.37, stdev=31.22
    clat (usec): min=56, max=1117.2k, avg=424.38, stdev=9141.95
     lat (usec): min=61, max=1117.2k, avg=432.85, stdev=9143.92
    clat percentiles (usec):
     |  1.00th=[   100],  5.00th=[   176], 10.00th=[   184], 20.00th=[   188],
     | 30.00th=[   192], 40.00th=[   194], 50.00th=[   196], 60.00th=[   200],
     | 70.00th=[   208], 80.00th=[   221], 90.00th=[   262], 95.00th=[   330],
     | 99.00th=[  1090], 99.50th=[  4015], 99.90th=[ 17433], 99.95th=[ 55313],
     | 99.99th=[413139]
   bw (  KiB/s): min=   32, max=63288, per=100.00%, avg=39150.46, stdev=20516.18, samples=52
   iops        : min=    8, max=15822, avg=9787.62, stdev=5129.05, samples=52
  lat (usec)   : 100=1.01%, 250=86.51%, 500=9.48%, 750=1.05%, 1000=0.80%
  lat (msec)   : 2=0.46%, 4=0.19%, 10=0.35%, 20=0.05%, 50=0.04%
  lat (msec)   : 100=0.02%, 250=0.01%, 500=0.02%, 750=0.01%, 1000=0.01%
  lat (msec)   : 2000=0.01%
  cpu          : usr=1.50%, sys=7.48%, ctx=517800, majf=0, minf=1
  IO depths    : 1=0.1%, 2=0.1%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=0,262144,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=4

Run status group 0 (all jobs):
  WRITE: bw=36.0MiB/s (37.8MB/s), 36.0MiB/s-36.0MiB/s (37.8MB/s-37.8MB/s), io=1024MiB (1074MB), run=28415-28415msec

Disk stats (read/write):
  sdb: ios=5/262962, merge=2/188, ticks=114/114428, in_queue=125645, util=73.82%
❰bhzhu❙~❱✔≻ fio -ioengine=libaio -bs=4k  -thread -rw=randwrite -size=1G -filename=test1 -name="Max throughput" -iodepth=4 -runtime=60
Max throughput: (g=0): rw=randwrite, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=4
fio-3.28
Starting 1 thread
Jobs: 1 (f=0): [f(1)][-.-%][w=118MiB/s][w=30.2k IOPS][eta 00m:00s] 
Max throughput: (groupid=0, jobs=1): err= 0: pid=200287: Fri Mar 18 19:26:45 2022
  write: IOPS=67.9k, BW=265MiB/s (278MB/s)(1024MiB/3861msec); 0 zone resets
    slat (usec): min=3, max=30327, avg=13.85, stdev=315.86
    clat (usec): min=2, max=30356, avg=44.55, stdev=547.61
     lat (usec): min=11, max=30361, avg=58.49, stdev=632.33
    clat percentiles (usec):
     |  1.00th=[   13],  5.00th=[   13], 10.00th=[   13], 20.00th=[   13],
     | 30.00th=[   13], 40.00th=[   14], 50.00th=[   14], 60.00th=[   15],
     | 70.00th=[   15], 80.00th=[   20], 90.00th=[   34], 95.00th=[   47],
     | 99.00th=[   58], 99.50th=[   71], 99.90th=[10552], 99.95th=[12518],
     | 99.99th=[14615]
   bw (  KiB/s): min=74320, max=858352, per=100.00%, avg=287481.14, stdev=332176.75, samples=7
   iops        : min=18580, max=214588, avg=71870.29, stdev=83044.19, samples=7
  lat (usec)   : 4=0.01%, 20=82.41%, 50=14.53%, 100=2.78%, 250=0.01%
  lat (usec)   : 500=0.01%, 750=0.02%
  lat (msec)   : 10=0.08%, 20=0.17%, 50=0.01%
  cpu          : usr=4.97%, sys=34.92%, ctx=258, majf=0, minf=1
  IO depths    : 1=0.1%, 2=0.1%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=0,262144,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=4

Run status group 0 (all jobs):
  WRITE: bw=265MiB/s (278MB/s), 265MiB/s-265MiB/s (278MB/s-278MB/s), io=1024MiB (1074MB), run=3861-3861msec

Disk stats (read/write):
  sdb: ios=0/5584, merge=0/12895, ticks=0/21299, in_queue=22673, util=69.60%
❰bhzhu❙~❱✔≻ fio -ioengine=libaio -bs=4k  -thread -rw=randread -size=1G -filename=test1 -name="Max throughput" -iodepth=4 -runtime=60
Max throughput: (g=0): rw=randread, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=4
fio-3.28
Starting 1 thread
Jobs: 1 (f=1): [r(1)][100.0%][r=800KiB/s][r=200 IOPS][eta 00m:00s]
Max throughput: (groupid=0, jobs=1): err= 0: pid=200350: Fri Mar 18 19:27:58 2022
  read: IOPS=182, BW=729KiB/s (746kB/s)(42.7MiB/60002msec)
    slat (nsec): min=1802, max=87492k, avg=5257829.30, stdev=3913061.47
    clat (usec): min=5, max=123221, avg=15792.04, stdev=7390.41
     lat (usec): min=26, max=128076, avg=21050.55, stdev=8646.24
    clat percentiles (msec):
     |  1.00th=[    3],  5.00th=[    7], 10.00th=[    9], 20.00th=[   11],
     | 30.00th=[   13], 40.00th=[   14], 50.00th=[   16], 60.00th=[   17],
     | 70.00th=[   19], 80.00th=[   21], 90.00th=[   23], 95.00th=[   25],
     | 99.00th=[   39], 99.50th=[   52], 99.90th=[   87], 99.95th=[  100],
     | 99.99th=[  120]
   bw (  KiB/s): min=   48, max=  944, per=100.00%, avg=753.95, stdev=94.77, samples=115
   iops        : min=   12, max=  236, avg=188.49, stdev=23.69, samples=115
  lat (usec)   : 10=0.01%, 20=0.02%, 50=0.08%, 250=0.05%, 500=0.27%
  lat (usec)   : 750=0.05%, 1000=0.02%
  lat (msec)   : 2=0.10%, 4=1.52%, 10=15.37%, 20=60.82%, 50=21.17%
  lat (msec)   : 100=0.48%, 250=0.05%
  cpu          : usr=0.14%, sys=1.94%, ctx=20048, majf=0, minf=6
  IO depths    : 1=0.1%, 2=0.1%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=10933,0,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=4

Run status group 0 (all jobs):
   READ: bw=729KiB/s (746kB/s), 729KiB/s-729KiB/s (746kB/s-746kB/s), io=42.7MiB (44.8MB), run=60002-60002msec

Disk stats (read/write):
  sdb: ios=10165/8880, merge=0/9121, ticks=56276/27758, in_queue=87532, util=99.08%
❰bhzhu❙~❱✔≻ 







fio -ioengine=libaio -bs=4k  -thread -rw=read -size=1G -filename=test1 -name="Max throughput" -iodepth=4 -runtime=60
Max throughput: (g=0): rw=read, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=4
fio-3.28
Starting 1 thread
Jobs: 1 (f=1): [R(1)][100.0%][r=179MiB/s][r=45.8k IOPS][eta 00m:00s]
Max throughput: (groupid=0, jobs=1): err= 0: pid=202768: Fri Mar 18 19:58:32 2022
  read: IOPS=36.9k, BW=144MiB/s (151MB/s)(1024MiB/7109msec)
    slat (nsec): min=725, max=93793k, avg=25457.61, stdev=412854.10
    clat (usec): min=2, max=93843, avg=81.94, stdev=714.30
     lat (usec): min=4, max=94074, avg=107.52, stdev=825.93
    clat percentiles (usec):
     |  1.00th=[    5],  5.00th=[    8], 10.00th=[   13], 20.00th=[   14],
     | 30.00th=[   14], 40.00th=[   15], 50.00th=[   15], 60.00th=[   16],
     | 70.00th=[   17], 80.00th=[   23], 90.00th=[  194], 95.00th=[  281],
     | 99.00th=[  529], 99.50th=[  783], 99.90th=[10421], 99.95th=[16057],
     | 99.99th=[23462]
   bw (  KiB/s): min=60984, max=216040, per=100.00%, avg=147504.57, stdev=46143.10, samples=14
   iops        : min=15246, max=54010, avg=36876.14, stdev=11535.78, samples=14
  lat (usec)   : 4=0.03%, 10=6.81%, 20=70.86%, 50=6.56%, 100=1.59%
  lat (usec)   : 250=7.05%, 500=6.02%, 750=0.54%, 1000=0.16%
  lat (msec)   : 2=0.06%, 4=0.06%, 10=0.16%, 20=0.08%, 50=0.03%
  lat (msec)   : 100=0.01%
  cpu          : usr=5.66%, sys=29.70%, ctx=14586, majf=0, minf=6
  IO depths    : 1=0.1%, 2=0.1%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=262144,0,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=4

Run status group 0 (all jobs):
   READ: bw=144MiB/s (151MB/s), 144MiB/s-144MiB/s (151MB/s-151MB/s), io=1024MiB (1074MB), run=7109-7109msec

Disk stats (read/write):
  sdb: ios=16700/25, merge=32503/12, ticks=8228/275, in_queue=8745, util=98.53%
❰bhzhu❙~❱✔≻ fio -ioengine=libaio -bs=4k -direct=1 -thread -rw=write -size=1G -filename=test1 -name="Max throughput" -iodepth=4 -runtime=60
Max throughput: (g=0): rw=write, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=4
fio-3.28
Starting 1 thread
Jobs: 1 (f=1): [W(1)][100.0%][w=47.5MiB/s][w=12.1k IOPS][eta 00m:00s]
Max throughput: (groupid=0, jobs=1): err= 0: pid=202851: Fri Mar 18 19:59:39 2022
  write: IOPS=10.9k, BW=42.7MiB/s (44.8MB/s)(1024MiB/23973msec); 0 zone resets
    slat (usec): min=2, max=13434, avg= 8.32, stdev=27.28
    clat (usec): min=52, max=568034, avg=353.79, stdev=5538.03
     lat (usec): min=57, max=568038, avg=362.20, stdev=5538.07
    clat percentiles (usec):
     |  1.00th=[    94],  5.00th=[   178], 10.00th=[   184], 20.00th=[   188],
     | 30.00th=[   190], 40.00th=[   194], 50.00th=[   196], 60.00th=[   200],
     | 70.00th=[   208], 80.00th=[   217], 90.00th=[   251], 95.00th=[   297],
     | 99.00th=[  1004], 99.50th=[  3359], 99.90th=[  9634], 99.95th=[ 23725],
     | 99.99th=[291505]
   bw (  KiB/s): min=18184, max=61160, per=99.82%, avg=43662.30, stdev=8692.43, samples=47
   iops        : min= 4546, max=15290, avg=10915.57, stdev=2173.11, samples=47
  lat (usec)   : 100=1.17%, 250=88.66%, 500=7.34%, 750=1.01%, 1000=0.82%
  lat (msec)   : 2=0.39%, 4=0.14%, 10=0.38%, 20=0.04%, 50=0.02%
  lat (msec)   : 100=0.01%, 250=0.01%, 500=0.02%, 750=0.01%
  cpu          : usr=1.57%, sys=9.81%, ctx=520303, majf=0, minf=2
  IO depths    : 1=0.1%, 2=0.1%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=0,262144,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=4

Run status group 0 (all jobs):
  WRITE: bw=42.7MiB/s (44.8MB/s), 42.7MiB/s-42.7MiB/s (44.8MB/s-44.8MB/s), io=1024MiB (1074MB), run=23973-23973msec

Disk stats (read/write):
  sdb: ios=5/262007, merge=0/130, ticks=30/79661, in_queue=86735, util=98.29%

```

**Comment by kagiannis** [comment]
Judging by the results you show you have an HDD. In HDDs random reads or writes are slow. Bcachefs is a COW filesystem in which writes are written sequentially. This is why random writes are very fast but random reads cause long seek times. I think this is the expected speed.

**Comment by YellowOnion** [comment]
What's your output of `bcachefs fs usage`?

**Comment by bhzhu203** [comment]
> bcachefs fs usage

sudo bcachefs fs usage /home
Filesystem 0b94d0c5-97da-4798-9c51-c764341cb8de:
Size:               398969704448
Used:               193759535104
Online reserved:         7565312

Data type       Required/total  Devices
reserved:       1/0             []                              14868480
btree:          1/1             [sdb9]                           2646081536
user:           1/1             [sdb9]                           188939777536

(no label) (device 0):                          sdb9              rw
                                data         buckets      fragmented
  sb:                        3149824              13          258048
  journal:                2147483648            8192               0
  btree:                  2646081536           10094               0
  user:                 188939777536          757035      9512405504
  cached:                          0               0               0
  parity:                          0               0               0
  erasure coded:                   0               0               0
  available:            230413565952          878958
  capacity:             433662722048         1654292
